File: abstractions/ruby/microsoft_kiota_abstractions/lib/microsoft_kiota_abstractions/authentication/base_bearer_token_authentication_provider.rb
# frozen_string_literal: true

require_relative './authentication_provider'
require_relative './access_token_provider'

module MicrosoftKiotaAbstractions
  # Provides a base class for implementing AuthenticationProvider for Bearer token scheme
  class BaseBearerTokenAuthenticationProvider
    include MicrosoftKiotaAbstractions::AuthenticationProvider
    def initialize(access_token_provider)
      raise StandardError, 'access_token_provider parameter cannot be nil' if access_token_provider.nil?

      @access_token_provider = access_token_provider
    end 

    AUTHORIZATION_HEADER_KEY = 'Authorization'
    def authenticate_request(request, additional_properties = {})
      raise StandardError, 'Request cannot be null' if request.nil?

      Fiber.new do
        token = @access_token_provider.get_authorization_token(request.uri, additional_properties).resume
        request.headers.add(AUTHORIZATION_HEADER_KEY, "Bearer #{token}") unless token.nil? || token.empty?
      end unless request.headers.get_all.key?(AUTHORIZATION_HEADER_KEY)
    end
  end
end

File: abstractions/python/kiota/abstractions/authentication/base_bearer_token_authentication_provider.py
from ..request_information import RequestInformation
from .access_token_provider import AccessTokenProvider
from .authentication_provider import AuthenticationProvider


class BaseBearerTokenAuthenticationProvider(AuthenticationProvider):
    """Provides a base class for implementing AuthenticationProvider for Bearer token scheme.
    """
    AUTHORIZATION_HEADER = "Authorization"

    def __init__(self, access_token_provider: AccessTokenProvider) -> None:
        self.access_token_provider = access_token_provider

    async def authenticate_request(self, request: RequestInformation) -> None:
        """Authenticates the provided RequestInformation instance using the provided
        authorization token

        Args:
            request (RequestInformation): Request information object
        """
        if not request:
            raise Exception("Request cannot be null")
        if not request.request_headers:
            request.headers = {}

        if not self.AUTHORIZATION_HEADER in request.headers:
            token = await self.access_token_provider.get_authorization_token(request.url)
            if token:
                request.add_request_headers({f'{self.AUTHORIZATION_HEADER}': f'Bearer {token}'})
